File: src/main/javabak/block/TieredModBlockItem.java
package redcrafter07.processed.block;

import net.minecraft.network.chat.Component;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import org.jetbrains.annotations.NotNull;
import redcrafter07.processed.block.machine_abstractions.TieredProcessedBlock;

import java.util.List;

public class TieredModBlockItem extends
        BlockItem {
    private final TieredProcessedBlock tieredProcessedBlock;

    public TieredModBlockItem(TieredProcessedBlock tieredProcessedBlock, Properties itemProperties) {
        super(tieredProcessedBlock, itemProperties);
        this.tieredProcessedBlock = tieredProcessedBlock;
    }

    @Override
    public void appendHoverText(
            @NotNull ItemStack stack,
            @NotNull TooltipContext context,
            @NotNull List<Component> tooltip,
            @NotNull TooltipFlag flag
    ) {
        tieredProcessedBlock.getDescription(tooltip, flag);

        if (getBlock() instanceof AdditionalBlockInfo block) {
            Component additionalTooltip = block.getAdditionalTooltip(stack, context, flag);
            if (additionalTooltip != null) tooltip.add(additionalTooltip);
        }

        tooltip.add(Component.empty());
        super.appendHoverText(stack, context, tooltip, flag);
    }

    @Override
    public @NotNull Component getName(@NotNull ItemStack itemStack) {
        return tieredProcessedBlock.getName();
    }
}

File: src/main/kotlin/redcrafter07/processed/block/TieredModBlockItem.kt
package redcrafter07.processed.block

import net.minecraft.network.chat.Component
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.level.block.Block

class TieredModBlockItem<T>(val tieredBlock: T, itemProperties: Properties) :
    BlockItem(tieredBlock, itemProperties) where T : Block, T : TieredBlock {
    override fun appendHoverText(
        stack: ItemStack, context: TooltipContext, tooltip: MutableList<Component>, flag: TooltipFlag
    ) {
        tieredBlock.getDescription(tooltip, flag)

        val block = tieredBlock
        if (block is AdditionalBlockInfo) {
            val additionalTooltip: Component? = block.getAdditionalTooltip(stack, context, flag)
            if (additionalTooltip != null) tooltip.add(additionalTooltip)
        }

        tooltip.add(Component.empty())
        super.appendHoverText(stack, context, tooltip, flag)
    }

    override fun getName(itemStack: ItemStack): Component {
        return tieredBlock.name
    }
}